### DISCUSSION TITLE
How to use LR11x0 examples in Arduino IDE?

### DISCUSSION THREAD

**Discussion by K4KDR** [jgromes/RadioLib#1101] [Q&A]
Greetings!  I've searched online until I can't search any more, so any help appreciated.

I'm very anxious to learn more about **LR-FHSS** which, as best as I can find here in RadioLib, so far only is referenced in examples for the **LR11x0** hardware.

`https://github.com/jgromes/RadioLib/tree/master/examples/LR11x0`

Wanting to use the Arduino IDE (since that's where I do everything else w/ RadioLib), I thought it would be a simple matter to Google search & find a URL to append to the **Board Manager** locations to be searched in the Arduino IDE.   Wrong - I've been unable to find a single reference to how any of the LR11x0 devices can be programmed in the Arduino IDE.  

As for hardware, in the interest of not spending TOO much on something I may or may not end up using, I have the following on the way from Mouser:

https://www.semtech.com/products/wireless-rf/lora-edge/lr1110mb1lcks

... but I'll be the first to admit that I have absolutely NO idea what I'm going to do with it when it arrives.   I've got an endless collection of LoRa shields for the Raspberry Pi, Arduino Uno, and Pi Pico platforms, but don't have a clue where to start w/ this Semtech LR1110 kit.  But you all know how that goes... you have to start somewhere & we learn as we go.  I'm very comfortable with all those other platforms (thanks to a lot of help from Jan & others), so I feel pretty dumb to be so clueless on this new-to-me environment.

Anyway, clearly I am COMPLETELY missing some key element of how I'll be able to use the Arduino IDE to integrate the RadioLib LR-FHSS example code into other, fully functional RadioLib examples (as instructed in the code - I know you don't try to run it exactly as-is)... and then run it on this Semtech LR1110 kit.

What am I missing?  Or, unlike most of the RadioLib code, can you NOT use the Arduino IDE for these LR11x0 devices?  If not, thanks in advance for any direction on what you DO use.  

-Scott, K4KDR

**Comment by jgromes** [comment]
Hi Scott!

> LR-FHSS (...) so far only is referenced in examples for the LR11x0 hardware.

Yes, that's correct - technically also the SX126x can perform LR-FHSS, but it can only transmit and it leaves building the packet (headers, FEC, whitening etc.) up to the user, so that hasn't been implemented yet.

> I have the following on the way from Mouser:

Looks to be pretty much just an Arduino shield, so I would start by finding a board that can take it and has a 3.3V logic (so no Arduino Uno). STM32 Nucleo kits usually work quite well and in my experience have very good documentation, including schematics. Then you just go to the Arduino Core for STM32 and get the boards definitions from there. On a side note, the LR11x0 is just an SPI device, much like the SX126x or SX127x, so integfacing with it is just a matter of hooking up the SPI and the BUSY, IRQ and RESET signals.

For my development, I have used Seeed WM1110 (https://www.seeedstudio.com/Wio-Tracker-1110-Dev-Board-p-5799.html) which integrates an nRF52 microcontroller with the LR1110 in a single package. The manufacturer does provide Arduino core for it, so I've been actually using Arduino IDE for running the examples throughout development.

Hope this helps!

**Reply by K4KDR** [reply]
>>> For my development, I have used Seeed WM1110 (https://www.seeedstudio.com/Wio-Tracker-1110-Dev-Board-p-5799.html)

After having so much success transmitting properly formatted LoRaWAN packets using the **SemTech LR1110 shield** (https://www.semtech.com/products/wireless-rf/lora-edge/lr1110mb1lcks) sitting on top of the **Nucleo-L476RG board** (https://www.digikey.com/en/products/detail/stmicroelectronics/NUCLEO-L476RG/5347711), I moved on to try to repeat that success w/ the 

`radio.beginLRFHSS();`

method.   Running into errors, I decided to take a step back and also get familiar with using the **Seeed WM1110** since you mentioned using that yourself.  I wouldn't bother everyone here if I hadn't already spent hours & hours searching online, but I need to ask for help w/ the (NSS/CS, DIO1, RESET, BUSY) pin numbers for the SEEED device, please.  

I actually thought that I had possibly found the correct pin numbers picked from MULTIPLE sources one-by-one:

`LR1110 radio = new Module(44, 2, 18, 43);`

... but since my serial output just halts, I'm guessing these are not correct.   Could I trouble someone to please share the pin numbers that are working for you with the Seeed WM1110 hardware, please?

Many thanks!

**Reply by jgromes** [reply]
The pins I'm using are as follows - the SPI pins are not the defaults ones, so you need to assign them and start the SPI:

```c++
LR1110 radio = new Module(44, 2, 42, 43, SPI);
(...)
void setup() {
  SPI.setPins(47, 45, 46);
  SPI.begin();
  radio.begin();
```

**Reply by K4KDR** [reply]
Thanks so much!!

**Reply by K4KDR** [reply]
**Update:**   Not to leave any doubt, with the correct SPI commands now in-hand, the Seeed WM1110 Wio Tracker 1110 now transmits perfectly via a sketch based on the LoRaWAN_ABP example.   So now I have -2- devices that transmit properly using the radio.begin method.    Neither is YET working by substituting in the **radio.beginLRFHSS();** method from the LR11x0_LR_FHSS_Modem example, but it's on me to learn how to make that transition properly!  Thanks again for all the help

**Reply by jgromes** [reply]
Thanks for letting me know - full disclosure, I only have a single LR1110 device, so I was not able to fully test the LR-FHSS modem. I know that it does transmit something resembling FHSS, but it is possible the reception is broken.

**Reply by StevenCellist** [reply]
Also, LoRaWAN and LR-FHSS is a very Dark Arts area. I have only recently learned that it exists, but also learned that it is hardly supported and there is nothing in RadioLib's LoRaWAN stack that does something with LR-FHSS. LoRaMac-node might support it, but that is a heck of a mess, and you would also need a very expensive gateway to receive LoRaWAN LR-FHSS. 
P2P LoRa over LR-FHSS might work just fine, that's not my expertise.

**Comment by K4KDR** [comment]
Thanks a million for the reply!!   Naturally you were able to direct me to a less expensive LR11x0 device, but that's ok; the SemTech kit wasn't terrible.    Thanks for the voltage warning - I'm using the Pi-Pico microcontrollers for more and more these days, so maybe that can be used; I have several on-hand.    

Great news that I can stay in the IDE that I'm comfortable with.  Not being a 'real' programmer, any changes to my routine are an additional burden that I really don't need.    I'll look forward to getting familiar with another aspect of LoRa & LoRaWAN functionality.

Thanks!

**Comment by K4KDR** [comment]
After much effort, I continue to struggle with a RadioLib **version** situation that I don't understand.  Identified as v6.5, the version that automatically installs in the Arduino IDE does NOT include references to the LR11x0 modules.   So, I removed that folder and **manually cloned the current RadioLib MASTER branch** into my local Arduino/Libraries folder.  That also identifies as v6.5, but now includes module support for the LR11x0 devices as well as the associated EXAMPLES.     

My real problem, though, is not related to the LR1110 board that I acquired.  My problem now is that there seems to have been changes to the example that my work is based on:

`RadioLib/examples/LoRaWAN/LoRaWAN_ABP/LoRaWAN_ABP.ino`

... and using my proven good **Pi-Pico/sx1262** hardware, I can no longer compile without errors.  I was preparing a new, long issue w/ debug output, etc., (and will be glad to do that if needed), but wanted to follow-up here first to ask if I am just approaching the RadioLib library situation incorrectly.  Should I just wait for a new RadioLib version before trying to work with my LR1110 hardware?   Or am I just making the latest RadioLib library available to my local Arduino IDE incorrectly.  

For instance, using the latest (cloned) RadioLib, even though I DID notice that some variable names changed in the example I'm basing my work on (so I was careful to adjust my code to match the new names), I get:

```
In file included from sketch/configABP.h:4,
                 from /tmp/arduino_modified_sketch_928836/Pico_sx1262_LoRaWAN_new-RadioLib_EXAMPLE_K4KDR.ino:32:
/home/k4kdr/Arduino/libraries/RadioLib/src/RadioLib.h:56:32: note: '#pragma message: 
RadioLib Info
Version:  "6.5.0.0"
Platform: "Raspberry Pi Pico (unofficial)"
Compiled: "May 24 2024" "22:58:24"'
   56 |   #pragma message(RADIOLIB_INFO)
      |                                ^
/tmp/arduino_modified_sketch_928836/Pico_sx1262_LoRaWAN_new-RadioLib_EXAMPLE_K4KDR.ino: In function 'void setup()':
Pico_sx1262_LoRaWAN_new-RadioLib_EXAMPLE_K4KDR:47:24: error: no matching function for call to 'LoRaWANNode::beginABP(uint32_t&, uint8_t [16], uint8_t [16], uint8_t [16], uint8_t [16], bool)'
   47 |   state = node.beginABP(devAddr, fNwkSIntKey, sNwkSIntKey, nwkSEncKey, appSKey, true);
```

... has something changed w/ the "**beginABP**" command?  It works perfectly fine in my code using the RadioLib v6.5 that the Arduino IDE pulls in through the normal Library Manager.

Any guidance appreciated!     I'm very anxious to do more w/ my LR1110 hardware as it works fine - but I'm stuck between using the Arduino IDE RadioLib w/ NO support for LR11x0 modules and the issues mentioned above w/ the latest RadioLib cloned from github.  

Thanks so much!

**Reply by jgromes** [reply]
Just to clarify, the version you get from Arduino IDE is the latest [release](https://github.com/jgromes/RadioLib/releases). The LR11x0 driver has not been released yet, but it is in master. There has also been development on the LoRaWAN implementation, reworking the joining/activation logic. I'm aiming for a new release soon-ish ;)

**Comment by K4KDR** [comment]
Follow-up...

Against all odds, and with a huge thanks to those of you maintaining this code, I was able to adapt my LR1110 ABP sketch to properly utilize the latest code here in RadioLib!

If anyone else experimenting with an LR1110 module is interested, I will list the details.  Please be aware that I still may not be using this code correctly since I'm a novice and don't fully understand all the syntax.

It's important to mention that before attempting to use ANY of the latest RadioLib code w/ my new LR1110 board, I ran the **firmware update** example:

`RadioLib/examples/LR11x0/LR11x0_Firmware_Update/LR11x0_Firmware_Update.ino`

... with the following result:

```
[LR1110] Initializing ... success!
01:16:25.864 -> [LR1110] Reading firmware versions ... success!
01:16:25.930 -> [LR1110] Device: 1
01:16:25.930 -> [LR1110] Base firmware: 3.3
01:16:25.963 -> [LR1110] WiFi firmware: 2.1
01:16:25.996 -> [LR1110] GNSS firmware: 0.0
01:16:26.029 -> [LR1110] Send any character to start the update
01:16:40.213 -> [LR1110] Updating firmware, this may take several seconds ... success!
01:16:48.931 -> [LR1110] Reading firmware versions ... success!
01:16:48.998 -> [LR1110] Device: 1
01:16:49.030 -> [LR1110] Base firmware: 4.1
01:16:49.063 -> [LR1110] WiFi firmware: 3.2
01:16:49.063 -> [LR1110] GNSS firmware: 2.0
```

... and now on to my main project which is based on the example:

`RadioLib/examples/LoRaWAN/LoRaWAN_ABP/LoRaWAN_ABP.ino`

... first, the pin numbers used in the **configABP.h** file for communications to my LR1110 board thru the nucleo L476RG board are set by:

```
#define Semtech_LR1110_expansion_board   // set to match hardware in-use (see pin maps below!)
~~~~~
// LR1110 module on nucleo_l476rg board
#elif defined(Semtech_LR1110_expansion_board)
  LR1110 radio = new Module(D7, D5, A0, D3);  // pin order: Module(NSS/CS, DIO1, RESET, BUSY)
```

... and in the main .INO file, I used:

```
  node.beginABP(devAddr, NwkSEncKey, SNwkSIntKey, NwkSKey, AppSKey);
  node.activateABP();
```

... and when the LR1110 transmitted my 5-byte payload over-the-air, the resulting 18-byte packet was received by another LoRa device that was monitoring the frequency that I was using:

![arduino-raw-RX](https://github.com/jgromes/RadioLib/assets/19886368/cf368a75-f994-4980-b17b-2c5adad8241d)

... but of course it's worthless if my original 5-byte payload ("K4KDR") was not actually in the encrypted LoRaWAN packet.  To my amazement, just as I'd previously accomplished w/  Pi-Pico/sx1262 hardware, **it was now working with my new LR1110 board**.

![LR1110_good_decode](https://github.com/jgromes/RadioLib/assets/19886368/dfc03c22-5b69-40ed-a3e8-dde5ffc06368)

... NOW, I can get to the point of the whole LR1110 purchase - to see what I can learn about **LR-FHSS**!

Thanks again to everyone.   I'll monitor the repository for future changes and adjust my code as required to stay current with the standards that you all settle on.

**Comment by K4KDR** [comment]
Ok, first success with LR-FHSS!

I'll ask a question up front instead of waiting until the end, since LF-FHSS is so new and I was only able to get it working - not completely error free - with one method (**radio.transmit**), would you like for me to create an ISSUE detailing the specific errors that I encountered along w/ the debug logs from each attempt?   Thank you.

As for this first partial success, my end goal is to have functional LoRaWAN / ABP transmit capability.  However, here on day-1 I ran into a multitude of errors that, thanks to the debug log, I was able to trace to the radio settings that the sketch was attempting to use from the US915 bandplan via 

`const LoRaWANBand_t Region = US915;`

... and 

`LoRaWANNode node(&radio, &Region, subBand);`

... as well as further errors when trying to transmit with

`int state = node.sendReceive(uplinkPayload, sizeof(uplinkPayload));`

... so, it was a simple matter to instead use the more basic command suggested in the example:

`int state = radio.transmit`   

... with a couple dozen characters to cause the transmitted payload to last more than a fraction of a second. (I commented out all the LoRaWAN commands in my sketch just to see if I could generate ANY LR-FHSS)

On an SDR centered on the frequency specified by radio.setFrequency , that resulted in something quite different from all the LoRa packets I have seen to date!

![1st_FHSS_TX_radio transmit](https://github.com/jgromes/RadioLib/assets/19886368/4bc0baa9-07dc-423a-91cc-30e616bfa729)

... so, I'm quite thrilled to break the ice on LR-FHSS.  Thank  you all for the help!  (you can see in the serial terminal that the transmit command actually times out... but clearly it DOES transmit over-the air)

I should briefly mention why I am so interested in LR-FHSS... during a recent presentation I heard someone from one of the companies trying to establish satellite-based IoT/LoRaWAN relay in orbit very casually say,   "... and OF COURSE, all of the uplinks to our satellites are LR-FHSS".     While there often isn't a lot of technical info in these online presentations, that comment stuck with me and made me want to learn as much as I could about LR-FHSS.  So I'm very grateful to all of you for getting me this far.

If there is any interest in me creating an ISSUE to serve as a home base for any further development of LR-FHSS in RadioLib, please let me know.   FYI, my test hardware is:

=Pi-Pico w/ sx1262 shield
=Nucleo L476RG w/ Semtech LR1110 development board
=Seeed WM1110 Wio Tracker 1110 (LoRa radio built-in)

**Reply by jgromes** [reply]
I seem to have missed a crucial piece of information - that you intend to use LR-FHSS in LoRaWAN. To be honest, I didn't know that was supported, but a quick glance into the latest LoRaWAN regional parameters (RP002-1.0.4) shows that it indeed is the case.

I any case, RadioLib's LoRaWAN stack does not support LR-FHSS, though I think it could be added. I will need to investigate whether my testing gateway is capable of receiving LR-FHSS, and how to configure the network settings for this case.

> On an SDR centered on the frequency specified by radio.setFrequency , that resulted in something quite different from all the LoRa packets I have seen to date!

That indeed looks like LR-FHSS. It's a completely different modulation than LoRa, so naturally it looks quite different in the waterfall.

**Reply by K4KDR** [reply]
Thanks for replying!   Yes - I've been studying the same regional parameters document to familiarize myself with the possible LR-FHSS settings that comply.  Indeed, there is currently no exact configuration match in RadioLib (which of course is no surprise since this is all new).   Fortunately, at least using **US915** as an example, the list of potential settings is small... only two data rates:

```
LR-FHSS 1.523 MHz BW at DR5-DR6, 
----------------------------------
DR5: LR-FHSS CR1/3: 1.523 MHz BW
DR6: LR-FHSS CR2/3: 1.523 MHz BW
----------------------------------
```

![LR-FHSS_physical_layer_settings](https://github.com/jgromes/RadioLib/assets/19886368/85d8de6c-2eb1-4539-82e2-3ea91b0defc8)



... so, I know it takes a lot of time & effort (not to mention there are multiple regions to provide settings for), but if you decide to develop LoRaWAN LR-FHSS support, it'll be much appreciated.   Would seem it's going to become the norm for IoT/LoRaWAN satellite relay comms.

Thanks so much!

**Reply by StevenCellist** [reply]
LoRaWAN gateways with LR-FHSS are rare, and as per the TTN forum may well break the four-figure barrier. Also, network servers may not even provide LR-FHSS support even though the specs mention that they are available.
I am personally interested in LR-FHSS as well, but to get it implemented, there are multiple barriers:
- get a LR-FHSS gateway
- find out which network server supports it
- implement LoRaWAN v1.0.4 in RadioLib as LR-FHSS is not available in LoRaWAN v1.1 regional parameters
- finally, implement LR-FHSS into the RadioLib's LoRaWAN

If someone finds me time & money I'll be happy to help... but it's looking like a rough road.

**Reply by jgromes** [reply]
@K4KDR this was the only thing referencing LR-FHSS I was able to find in Semtech's SX1302 driver repository (which is the IC used in most LoRaWAN gateways): https://github.com/Lora-net/sx1302_hal/issues/110

So unfortunately it looks like the only way to test LoRaWAN over LR-FHSS is to either dish out several hundred bucks at the very least, or implement it yourself using some SDR.

**Reply by K4KDR** [reply]
Thanks so much for checking!

Since I have multiple devices, I have been attempting to use the RECEIVE portion of the example code in

`examples/LR11x0/LR11x0_LR_FHSS_Modem/LR11x0_LR_FHSS_Modem.ino`

... but so far, when my TX device sends out an LR-FHSS payload, the RX device just shows:

![Screenshot from 2024-05-31 06-02-04](https://github.com/jgromes/RadioLib/assets/19886368/10dd2d4f-ab73-4e27-88e8-0e43b0f6541d)

... so, while I'm very happy to be able to generate an LR-FHSS transmit payload, I have a lot more to learn about the RECEIVE part.

Thanks again!

**Reply by jgromes** [reply]
@K4KDR not sure if you're still interested in this, I just wanted to let you know that:

1. I have pushed a work-in-progress version of LR-FHSS support for the SX126x series in [this branch](https://github.com/jgromes/RadioLib/tree/sx126x-lr-fhss). However, I cannot test it because of point number 2.
2. The LR11x0 doesn't actually support LR-FHSS reception, it can only transmit, much like the SX126x (sigh ...):
![Screenshot_104](https://github.com/user-attachments/assets/acfc36a9-4a4f-457e-86bc-403a5276075c)

So the only way to receive LR-FHSS seem to be a LoRaWAN gateway that supports this modulation. Unfortunately those are rather tricky to come by.

**Reply by StevenCellist** [reply]
And for completeness sake: the same goes for LR1120 and LR1121.

**Reply by K4KDR** [reply]
Yes!   Very interested.    Thanks so much for the update.

**Comment by K4KDR** [comment]
Ok, another successful update.   Using the pin numbers previously found to work for the **Semtech LR1110 board on a Nucleo Nucleo L476RG**:

`LR1110 radio = new Module(D7, D5, A0, D3);`

... basic LR-FHSS transmit using `radio.beginLRFHSS();` and `radio.transmit("K4KDRK4KDRK4KDRK4KDRK4KDRK4KDR"); `produces the same spectrum that the Seeed WM1110 Wio Tracker 1110 device did w/ pin numbers & settings appropriate for that hardware.

![LR1110_Nucleo_LR-FHSS_test_waterfall](https://github.com/jgromes/RadioLib/assets/19886368/26c85459-6a68-4f0c-95aa-3f5db5685336)


Thanks for all the help!

**Comment by K4KDR** [comment]
... and, the exact same sketch & hardware if you change from

` radio.beginLRFHSS();`

to

` radio.begin();`

![LR1110_Nucleo_regular_LoRa_test_waterfall](https://github.com/jgromes/RadioLib/assets/19886368/26f13e95-b018-439b-8f9b-2a36b3620dc7)


... always fun to see a comparison - this second waterfall of course a regular LoRa packet that has been around for quite a while now.
